dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/RequestItemRepository.java

```diff
@@ -0,0 +1,146 @@
+public class RequestItemRepository
+        extends DSpaceRestRepository<RequestItemRest, String> {
+    @Autowired(required = true)
+    protected RequestItemService requestItemService;
+
+    @Autowired(required = true)
+    protected BitstreamService bitstreamService;
+
+    @Autowired(required = true)
+    protected ItemService itemService;
+
+    @Autowired(required = true)
+    protected RequestItemConverter requestItemConverter;
+
+    @Override
+    @PreAuthorize("permitAll()")
+    public RequestItemRest findOne(Context context, String id) {
+        RequestItem requestItem = requestItemService.findByToken(context, id);
+        if (null == requestItem) {
+            return null;
+        } else {
+            return requestItemConverter.convert(requestItem, new DefaultProjection());
+        }
+    }
+
+    @Override
+    public Page<RequestItemRest> findAll(Context context, Pageable pageable) {
+        throw new RepositoryMethodNotImplementedException(RequestItemRest.NAME, "findAll");
+    }
+
+    @Override
+    @PreAuthorize("permitAll()")
+    public RequestItemRest createAndReturn(Context ctx) {
+        // Fill a RequestItemRest from the client's HTTP request.
+        HttpServletRequest req = getRequestService()
+                .getCurrentRequest()
+                .getHttpServletRequest();
+        ObjectMapper mapper = new ObjectMapper();
+        RequestItemRest rir;
+        try {
+            rir = mapper.readValue(req.getInputStream(), RequestItemRest.class);
+        } catch (IOException ex) {
+            throw new UnprocessableEntityException("error parsing the body", ex);
+        }
+
+        // Create the item request model object from the REST object.
+        String token;
+        try {
+            String bitstreamId = rir.getBitstreamId();
+            if (isBlank(bitstreamId)) {
+                throw new IncompleteItemRequestException("A bitstream ID is required");
+            }
+            Bitstream bitstream = bitstreamService.find(ctx, UUID.fromString(bitstreamId));
+            if (null == bitstream) {
+                throw new IncompleteItemRequestException("That bitstream does not exist");
+            }
+
+            String itemId = rir.getItemId();
+            if (isBlank(itemId)) {
+                throw new IncompleteItemRequestException("An item ID is required");
+            }
+            Item item = itemService.find(ctx, UUID.fromString(itemId));
+            if (null == item) {
+                throw new IncompleteItemRequestException("That item does not exist");
+            }
+
+            boolean allFiles = rir.isAllfiles();
+
+            String email = rir.getRequestEmail();
+            if (isBlank(email)) {
+                throw new IncompleteItemRequestException("A submitter's email address is required");
+            }
+
+            String username = rir.getRequestName();
+            String message = rir.getRequestMessage();
```

Comment:
Since these are user inputs, I think `message`, `username` and maybe even `email` should all be HTML encoded, to avoid any risk of an XSS style attack. I don't think any of these fields should support HTML/Javascript anyways, so best to encode or strip it before we even create the request.

(I guess this encoding could also be put into the `createRequest()` call below. Either way is fine.)

Reply:
Would you (or anyone) please point me toward an example.

Reply:
@mwoodiupui : it's likely sufficient to use one of the `StringEscapeUtils` methods here... likely `StringEscapeUtils.escapseHtml4()`: https://commons.apache.org/proper/commons-text/javadocs/api-release/org/apache/commons/text/StringEscapeUtils.html

This would just ensure that any attempts to insert HTML/JS in these fields would be turned instead into HTML entities, which should ensure they aren't rendered on any client.

Another option would be to see if we can use something like Apache Tika to *strip* the HTML out (I've never tried this, but seems possible): https://stackoverflow.com/a/12259801/3750035

UPDATE: Just realized we also already have a `Utils.addEntities()` method to do HTML escaping. So, you could just use that. https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/core/Utils.java#L270

Reply:
So, since I haven't found any other cases in dspace-server-webapp to study, let me see if I have this right:

- The contract should be updated to show that these fields are percent-encoded in flight.
- RequestItemRepository.createAndReturn must percent-DEcode them when creating the database row.
- RequestItemConverter.convert must percent-ENcode them when creating a RequestItemRest.

Reply:
@mwoodiupui : 
* The contract doesn't necessarily need to mention that the fields are encoded/decoded (it's up to you).  But, it DOES need to be updated to fully describe the endpoint. Currently, the contract is missing information about how to create a new Item Request.
* The end goal here is that we need to find a way to protect ourselves from XSS **because these endpoints allow for anonymous item request submissions**.   Ideally, the endpoint should NOT accept HTML at all -- it should only accept plain text.  So, this means either we strip any HTML tags...or we simply encode everything (which will invalidate HTML tags but potentially make the result look odd if someone attempts to submit an HTML formatted request).  It seems to me that we should never decode it, as doing so could result in the user reviewing the Item Request accidentally triggering an XSS attack (if the decoded request contains HTML/Javascript which then ends up being accidentally executed in the UI)

Does that help at all?

Reply:
So these fields should be HTML-encoded *in the database*?  That would be a change to the dspace-api layer and require fixup of existing requests during upgrade.

I really would like to look at some other code where we are already doing this.

Reply:
I'm beginning to wonder how this XSS would work.  Requests never appear in the UI.  Submission of the request form results in a text/plain email to the appropriate party.  This email contains a link which uses an opaque token to identify the request.  Following the link results in another text/plain email being returned to the requester, indicating acceptance or rejection and (in the case of acceptance) with the requested bitstream(s) attached.

I also begin to think that validating addresses and stripping markup would be more appropriate than this encoding that I still don't understand.

Reply:
@mwoodiupui : I sent a private response in Slack as well... but, my assumption was that Item Requests *do* get displayed in the UI in order to be reviewed/approved.  We had a ticket for that here: https://github.com/DSpace/dspace-angular/issues/779  You are correct that, if these requests _never appear in a UI_ then XSS isn't possible.  I had just been working on the assumption that the request needs to appear in the UI to be approved... in which case, we need to ensure that the Item Request cannot trigger an XSS attack (since that request can come from an anonymous user).

I admit, I am not very familiar with this feature overall...so, if there's no need for it to *ever appear* in the UI, then we can end this XSS discussion.  However, should we ever decide it needs a UI component (even to delete old requests or similar from an Admin), then we need to ensure safety from XSS at that time.

Reply:
I think that XSS has been addressed now.